src/datasets/iterable_dataset.py

```diff
@@ -2277,11 +2372,17 @@ def with_format(
         # TODO(QL): add format_kwargs
         # TODO(QL): add format_columns and return_all_columns
         # TODO(QL): add pandas format
+        ex_iterable = FormattedExamplesIterable(
+            self._ex_iterable,
+            formatting=FormattingConfig(format_type=type),
+            features=self.features,
+            token_per_repo_id=self._token_per_repo_id,
+        )
         return IterableDataset(
-            ex_iterable=self._ex_iterable,
+            ex_iterable=ex_iterable,
             info=self._info.copy(),
             split=self._split,
-            formatting=FormattingConfig(format_type=type),
+            formatting=None,  # formatting is applied in ex_iterable
```

Comment:
Originally we tried to keep `formatting` as an attribute of the `IterableDataset` so that people can change the formatting without having to create nested examples iterables that would apply a different formatting every time.

What is the rationale for removing it from here ?

Note that applying a formatting on data already in the right format is cheap.

Reply:
c.f. comment https://github.com/huggingface/datasets/pull/7207#discussion_r1800014191

my thinking was that this way the formatting is never `lost'; it only gets overridden if the user overrides it.

there could indeed be some nested cases that i didn't consider though, can you think of some?

Reply:
By nested I actually meant consecutive, I had in mind that this code would be a no-op:
```python
format = ds.format
ds = ds.with_format("np")
ds = ds.with_format(**format)
```

(actually just noticed `IterableDataset.format` is not implemented yet, though it should be and work like `Dataset.format`, but anyway the point is successive `with_format` would override the previous ones since it's an operation that is done "on top" of the dataset rather than a processing function)

Reply:
reverted this case so with_format.with_format can still no op